### PR TITLE
 Implement DOMPoint with magic dom

### PR DESCRIPTION
**Pull request by photoszzt**
I rebase the smup so it's pretty big... The real meat is the most recent one. 

r? @fitzgen

### CODE REVIEW

File: mozjs/js/rust/magic_codegen/src/lib.rs [hunk 1]
```diff
@@ -35,18 +37,32 @@ fn match_struct(ast: &syn::DeriveInput, variant: &syn::VariantData) -> quote::To
         .split('_')
         .next()
         .expect("Should have a '_' in the magic dom struct name");
+    let dom_flag = if name.starts_with("DOM") {
+        quote::Ident::from("jsapi::JSCLASS_IS_DOMJSCLASS")
```
Comment (fitzgen) [inline mozjs/js/rust/magic_codegen/src/lib.rs]: There are lots of types that are DOM classes that won't start with "DOM": `Node`, `Console`, etc.

Did you find it necessary to have this flag set? AFAICT, this is only used outside of the JS engine by embedders (gecko/servo) and I think we want to make whether it is set or not orthogonal to the magic dom stuff.

File: mozjs/js/rust/magic_codegen/src/lib.rs [hunk 2]
```diff
@@ -55,6 +71,16 @@ fn match_struct(ast: &syn::DeriveInput, variant: &syn::VariantData) -> quote::To
             #(#struct_def,)*
         }
 
+        pub static #js_class : jsapi::JSClass = jsapi::JSClass {
+            name: #name_field as *const u8 as *const libc::c_char,
+            flags: (#num_reserved_slots &
+                    jsapi::JSCLASS_RESERVED_SLOTS_MASK) << jsapi::JSCLASS_RESERVED_SLOTS_SHIFT |
+            jsapi::JSCLASS_HAS_PRIVATE  |
+            #dom_flag,
```
Comment (fitzgen) [inline mozjs/js/rust/magic_codegen/src/lib.rs]: If the name doesn't begin with `DOM`, then this is going to be an empty identifier, which will result in a syntax error right here, correct?

File: mozjs/js/rust/magic_codegen/src/lib.rs [hunk 3]
```diff
@@ -83,8 +121,38 @@ fn match_struct(ast: &syn::DeriveInput, variant: &syn::VariantData) -> quote::To
...
+                if !thisv.is_object() {
+                    return None;
+                }
+                let jsobj = thisv.to_object();
+                if jsapi::JS_GetClass(jsobj) as *const _ as usize != &#js_class as *const _ as usize {
+                    return None;
+                }
+                jsobj
+            };
+            Some(jsobj)
```
Comment (fitzgen) [inline mozjs/js/rust/magic_codegen/src/lib.rs]: This should probably be an unsafe function since it is using a raw pointer. I'd also tighten it up a little bit, like this:

```rust
pub unsafe fn check_this(cx: *mut jsapi::JSContext, args: &jsapi::JS::CallArgs) -> Option<#name> {
    let thisv = args.thisv().get();
    if !thisv.is_object() {
        return None;
    }
    #name::from_object(thisv.to_object())
}
```

I think this function could also make sense as a static method of `#name`, so it would be invoked as `#name::check_this(...)`

File: mozjs/js/rust/magic_codegen/src/lib.rs [hunk 4]
```diff
@@ -147,34 +212,197 @@ fn get_addr_test_struct_field(variant: &syn::VariantData) -> Vec<quote::Tokens>
     res
 }
 
-/// This function generates a tuple which consists the definition of the struct and the 
-/// definition of the slot number.
-fn get_struct_def(name: &quote::Ident, variant: &syn::VariantData) -> (Vec<quote::Tokens>, Vec<quote::Tokens>) {
+lazy_static! {
+    static ref JSVAL: std::collections::HashMap<syn::Ident, syn::Ident> = {
```
Comment (fitzgen) [inline mozjs/js/rust/magic_codegen/src/lib.rs]: A doc comment describing what this is and how it is used would be helpful for readers.

File: mozjs/js/rust/magic_codegen/src/lib.rs [hunk 5]
```diff
@@ -147,34 +212,197 @@ fn get_addr_test_struct_field(variant: &syn::VariantData) -> Vec<quote::Tokens>
...
+}
+
+/// This function generates a tuple which consists the definition of the struct, the
+/// definition of the slot number, getters and setters for the object, js accesible
+/// getter and setters and JSPropertySpec array elements.
+fn get_struct_def(name: &quote::Ident,
+                  variant: &syn::VariantData)
+                  -> (Vec<quote::Tokens>, Vec<quote::Tokens>, Vec<quote::Tokens>,
+                      Vec<quote::Tokens>, Vec<quote::Tokens>, Vec<quote::Tokens>,
+                      Vec<quote::Tokens>) {
```
Comment (fitzgen) [inline mozjs/js/rust/magic_codegen/src/lib.rs]: A couple nitpicks for readability/maintainability:

* Let's define a struct with field names instead of having to know the ordering of the tuples seven fields:

```rust
struct MagicStructCode {
    /// A definition of the magic struct itself.
    struct_def: Vec<quote::Tokens>,

    /// The definition of the number of slots the magic struct requires.
    slots: Vec<quote::Tokens>,

    /// Rust getters for the data stored in slots.
    getters: Vec<quote::Tokens>,

    /// Rust setters for the data stored in slots.
    setters: Vec<quote::Tokens>,

    /// `JSNative` getter functions exposed to JavaScript.
    js_getters: Vec<quote::Tokens>,

    /// `JSNative` setter functions exposed to JavaScript.
    js_setters: Vec<quote::Tokens>,

    /// The `JSPropertySpec` definition that exposes the `JSNative`s to JavaScript.
    js_prop_spec: Vec<quote::Tokens>,
}
```

* Can we pull generating each getter, setter, etc out into its own function that we call from inside the loop over each field? This function is a behemoth right now, which makes it hard to follow what is going on.

File: mozjs/js/rust/src/glue.rs [hunk 6]
```diff
@@ -339,6 +339,8 @@ extern "C" {
                                              -> bool;
 
     pub fn IsDebugBuild() -> bool;
+
+    // pub fn SimpleWarningReporter(cx: *mut JSContext, report: *mut JSErrorReport);
```
Comment (fitzgen) [inline mozjs/js/rust/src/glue.rs]: Let's remove this if it isn't used.

**Review by fitzgen — COMMENTED** [review]
Thanks @photoszzt! This looks great, I have just a few nitpicks below.

File: mozjs/js/rust/magic_codegen/src/lib.rs (hunk 1 above)
Comment (photoszzt) [reply mozjs/js/rust/magic_codegen/src/lib.rs]: I'm not sure it's necessary to set this flag. I just see the name and seems it may imply I need to add it.

Comment (fitzgen) [reply mozjs/js/rust/magic_codegen/src/lib.rs]: Let's skip it for now then, unless not having it leads to problems.

**Review by fitzgen — APPROVED** [review]
Thanks for fixing it up! Looks great!
